confignode/src/main/java/org/apache/iotdb/confignode/manager/PartitionManager.java

```diff
@@ -18,128 +18,197 @@
+    dataPartitionDataSet = (DataPartitionDataSet) consensusReadResponse.getDataset();
+    return dataPartitionDataSet;
   }
 
   /**
-   * TODO:allocate schema partition by balancer
+   * Get DataPartition and create a new one if it does not exist
    *
-   * @param physicalPlan physical plan
-   * @return data set
+   * @param physicalPlan DataPartitionPlan with Map<StorageGroupName, Map<SeriesPartitionSlot,
+   *     List<TimePartitionSlot>>>
+   * @return DataPartitionDataSet
    */
-  public DataSet applyDataPartition(DataPartitionPlan physicalPlan) {
-    return null;
+  public DataSet getOrCreateDataPartition(GetOrCreateDataPartitionPlan physicalPlan) {
+    Map<String, Map<SeriesPartitionSlot, List<TimePartitionSlot>>> noAssignedDataPartitionSlots =
+        partitionInfoPersistence.filterNoAssignedDataPartitionSlots(
+            physicalPlan.getPartitionSlotsMap());
+
+    if (noAssignedDataPartitionSlots.size() > 0) {
+      // Allocate DataPartition
+      Map<String, Map<SeriesPartitionSlot, Map<TimePartitionSlot, List<RegionReplicaSet>>>>
+          assignedDataPartition = allocateDataPartition(noAssignedDataPartitionSlots);
+      CreateDataPartitionPlan createPlan =
+          new CreateDataPartitionPlan(PhysicalPlanType.CreateDataPartition);
+      createPlan.setAssignedDataPartition(assignedDataPartition);
+
+      // Persistence DataPartition
+      ConsensusWriteResponse consensusWriteResponse = getConsensusManager().write(createPlan);
+      if (consensusWriteResponse.getStatus().getCode()
+          == TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+        LOGGER.info("Allocate data partition to {}.", assignedDataPartition);
+      }
+    }
+
+    return getDataPartition(physicalPlan);
   }
 
-  public DataSet getDataPartition(DataPartitionPlan physicalPlan) {
-    return null;
+  /**
+   * TODO: allocate by LoadManager
+   *
+   * @param noAssignedDataPartitionSlotsMap Map<StorageGroupName, Map<SeriesPartitionSlot,
+   *     List<TimePartitionSlot>>>
+   * @return assign result, Map<StorageGroupName, Map<SeriesPartitionSlot, Map<TimePartitionSlot,
+   *     List<RegionReplicaSet>>>>
+   */
+  private Map<String, Map<SeriesPartitionSlot, Map<TimePartitionSlot, List<RegionReplicaSet>>>>
+      allocateDataPartition(
+          Map<String, Map<SeriesPartitionSlot, List<TimePartitionSlot>>>
+              noAssignedDataPartitionSlotsMap) {
+
+    Map<String, Map<SeriesPartitionSlot, Map<TimePartitionSlot, List<RegionReplicaSet>>>> result =
+        new HashMap<>();
+
+    for (String storageGroup : noAssignedDataPartitionSlotsMap.keySet()) {
+      Map<SeriesPartitionSlot, List<TimePartitionSlot>> noAssignedPartitionSlotsMap =
+          noAssignedDataPartitionSlotsMap.get(storageGroup);
+      List<RegionReplicaSet> dataRegionEndPoints =
+          RegionInfoPersistence.getInstance().getDataRegionEndPoint(storageGroup);
+      Random random = new Random();
```

Comment:
make it a class field?

Reply:
The EndPoints of DataNodes are dynamic and maintained by DataNodeManager and DataNodeInfoPersistence. Therefore I suppose that there is not need to make EndPoints a class field in PartitionManager.

Reply:
I mean `Random`

Reply:
All interfaces associated with allocation in ConfigNode will be moved to LoadManager later(I marked TODO). Thus I suppose that make Random a class field is unnecessary.